016 - Koha API Biblios Patch
Patch against Koha $GITREF 9648cd2917f4b530c9da9b39e4fe606e457511a2
curl -s -o koha.tar.gz https://codeload.github.com/Koha-Community/Koha/legacy.tar.gz/9648cd2917f4b530c9da9b39e4fe606e457511a2
or https://github.com/Koha-Community/Koha/archive/v16.05.00-beta.tar.gz

prerequisites:
# LOCAL 015 - Koha API Items Patch

diff --git a/Koha/REST/V1/Biblio.pm b/Koha/REST/V1/Biblio.pm
new file mode 100644
index 0000000..06eb454
--- /dev/null
+++ b/Koha/REST/V1/Biblio.pm
@@ -0,0 +1,45 @@
+package Koha::REST::V1::Biblio;
+
+# This file is part of Koha.
+#
+# Koha is free software; you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation; either version 3 of the License, or (at your option) any later
+# version.
+#
+# Koha is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with Koha; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+use Modern::Perl;
+
+use Mojo::Base 'Mojolicious::Controller';
+
+use C4::Biblio qw( GetBiblioData );
+use Koha::Biblios;
+
+sub get {
+    my ($c, $args, $cb) = @_;
+
+    my $biblio = &GetBiblioData($args->{biblionumber});
+    unless ($biblio) {
+        return $c->$cb({error => "Biblio not found"}, 404);
+    }
+    return $c->$cb($biblio, 200);
+}
+
+sub getitems {
+    my ($c, $args, $cb) = @_;
+
+    my $biblio = Koha::Biblios->find($args->{biblionumber});
+    unless ($biblio) {
+        return $c->$cb({error => "Biblio not found"}, 404);
+    }
+    return $c->$cb({ biblio => $biblio->unblessed, items => $biblio->items->unblessed }, 200);
+}
+
+1;
diff --git a/api/v1/definitions/biblio.json b/api/v1/definitions/biblio.json
new file mode 100644
index 0000000..11f9578
--- /dev/null
+++ b/api/v1/definitions/biblio.json
@@ -0,0 +1,49 @@
+{
+  "type": "object",
+    "properties": {
+    "biblionumber": {
+      "type": ["string"],
+      "description": "Unique internal ID of biblio"
+    },
+    "author": {
+      "type": ["string", "null"],
+      "description": "statement of responsibility from MARC record (100$a in MARC21)"
+    },
+    "title": {
+      "type": ["string", "null"],
+      "description": "title (without the subtitle) from the MARC record (245$a in MARC21)"
+    },
+    "unititle": {
+      "type": ["string", "null"],
+      "description": "uniform title (without the subtitle) from the MARC record (240$a in MARC21)"
+    },
+    "notes": {
+      "type": ["string", "null"],
+      "description": "values from the general notes field in the MARC record (500$a in MARC21) split by bar (|)"
+    },
+    "serial": {
+      "type": ["string", "null"],
+      "description": "Boolean indicating whether biblio is for a serial"
+    },
+    "seriestitle": {
+      "type": ["string", "null"],
+      "description": ""
+    },
+    "copyrightdate": {
+      "type": ["string", "null"],
+      "description": "publication or copyright date from the MARC record"
+    },
+    "timestamp": {
+      "type": ["string", "null"],
+      "description": "date and time this record was last touched"
+    },
+    "datecreated": {
+      "type": ["string", "null"],
+      "description": "the date this record was added to Koha"
+    },
+    "abstract": {
+      "type": ["string", "null"],
+      "description": "summary from the MARC record (520$a in MARC21)"
+    }
+  }
+}
diff --git a/api/v1/definitions/biblios.json b/api/v1/definitions/biblios.json
new file mode 100644
index 0000000..908c03f
--- /dev/null
+++ b/api/v1/definitions/biblios.json
@@ -0,0 +1,5 @@
+{
+    "type": "array",
+    "items": { "$ref": "biblio.json" }
+}
+
diff --git a/api/v1/definitions/index.json b/api/v1/definitions/index.json
index 422eef2..2665d02 100644
--- a/api/v1/definitions/index.json
+++ b/api/v1/definitions/index.json
@@ -8,5 +8,7 @@
     "library": { "$ref": "library.json" },
     "items": { "$ref": "items.json" },
     "item": { "$ref": "item.json" },
+    "biblios": { "$ref": "biblios.json" },
+    "biblio": { "$ref": "biblio.json" },
     "error": { "$ref": "error.json" }
 }
diff --git a/api/v1/swagger.json b/api/v1/swagger.json
index 4659530..b2b0f4d 100644
--- a/api/v1/swagger.json
+++ b/api/v1/swagger.json
@@ -632,6 +632,66 @@
           }
         }
       }
+    },
+    "/biblios/{biblionumber}": {
+      "get": {
+        "operationId": "getBiblio",
+        "tags": ["biblios"],
+        "parameters": [
+          { "$ref": "#/parameters/biblionumberPathParam" }
+        ],
+        "produces": [
+          "application/json"
+        ],
+        "responses": {
+          "200": {
+            "description": "A biblio record",
+            "schema": { "$ref": "#/definitions/biblio" }
+
+          },
+          "403": {
+            "description": "Access forbidden",
+            "schema": {
+              "$ref": "#/definitions/error"
+            }
+          },
+          "404": {
+            "description": "Biblio not found",
+            "schema": {
+              "$ref": "#/definitions/error"
+            }
+          }
+        }
+      }
+    },
+    "/biblios/{biblionumber}/items": {
+      "get": {
+        "operationId": "getitemsByBiblio",
+        "tags": ["biblios", "items"],
+        "parameters": [
+          { "$ref": "#/parameters/biblionumberPathParam" }
+        ],
+        "produces": [
+          "application/json"
+        ],
+        "responses": {
+          "200": {
+            "description": "A biblio record with items"
+          },
+          "403": {
+            "description": "Access forbidden",
+            "schema": {
+              "$ref": "#/definitions/error"
+            }
+          },
+          "404": {
+            "description": "Biblio not found",
+            "schema": {
+              "$ref": "#/definitions/error"
+            }
+          }
+        }
+      }
     }
   },
   "definitions": {
@@ -665,6 +725,13 @@
       "description": "Internal item identifier",
       "required": true,
       "type": "integer"
+    },
+    "biblionumberPathParam": {
+      "name": "biblionumber",
+      "in": "path",
+      "description": "Internal biblio identifier",
+      "required": true,
+      "type": "integer"
     }
   }
 }